src/vicsetup/libvicsetup/dm.c

```diff
@@ -0,0 +1,917 @@
+#endif
+
+    /* Set the target */
+    if (!dm_task_add_target(dmt, start, size, "crypt", params))
+        RAISE(VIC_FAILED);
+
+#ifdef USE_UDEV
+    /* Determine whether sync support is available */
+    if ((sync = dm_udev_get_sync_support()))
+    {
+        uint16_t udev_flags = DM_UDEV_DISABLE_LIBRARY_FALLBACK;
+        dm_task_set_cookie(dmt, &cookie, udev_flags);
+    }
+#endif
+
+    /* Run the task to create the new target type */
+    if (!dm_task_run(dmt))
+        RAISE(VIC_FAILED);
+
+    /* Verify that the target now exists */
+    {
+        struct dm_info dmi;
+
+        if (!dm_task_get_info(dmt, &dmi) || !dmi.exists)
+            RAISE(VIC_FAILED);
+    }
+
+#ifdef USE_UDEV
+    if (sync)
+        dm_udev_wait(cookie);
+#endif
+
+done:
+
+    if (hexkey)
+        vic_free(hexkey);
+
+    if (dmt)
+    {
+        dm_task_destroy(dmt);
+        dm_task_update_nodes();
+    }
+
+    dm_lib_release();
+
+    return result;
+}
+
+vic_result_t vic_dm_create_integrity(
+    const char* name,
+    const char* path,
+    uint64_t start,
+    uint64_t size,
+    uint64_t offset,
+    char mode,
+    const char* integrity)
+{
+    vic_result_t result = VIC_OK;
+    char params[1024];
+    struct dm_task* dmt = NULL;
+    char* hexkey = NULL;
+    char dev[PATH_MAX];
+#ifdef USE_UDEV
+    int sync;
+    uint32_t cookie = 0;
+#endif
+
+    /* Reject invalid parameters */
+    if (!name || !path)
+        RAISE(VIC_BAD_PARAMETER);
+
+    /* If not a block device, then map to a loopback device */
+    {
+        struct stat st;
+
+        if (stat(path, &st) != 0)
+            RAISE(VIC_FAILED);
+
+        if (S_ISBLK(st.st_mode))
+        {
+            if (vic_strlcpy(dev, path, PATH_MAX) >= PATH_MAX)
+                RAISE(VIC_PATH_TOO_LONG);
+        }
+        else
+        {
+            if (vic_loop_attach(path, 0, false, false, dev) != 0)
+                RAISE(VIC_FAILED_TO_GET_LOOP_DEVICE);
+        }
+    }
+
+    /* Format the params */
+    {
+        size_t tag_size = vic_integrity_tag_size(integrity);
+
+        if (tag_size == (size_t)-1)
+            RAISE(VIC_UNEXPECTED);
+
+        /* ATTN-C: hard-coded block size */
+        uint64_t block_size = 512;
+
+        int n = snprintf(
+            params,
+            sizeof(params),
+            "%s %lu %lu %c 1 block_size:%lu",
+            dev,
+            offset,
+            tag_size,
+            mode,
+            block_size);
+
+        if (n <= 0 || (size_t)n >= sizeof(params))
+            RAISE(VIC_BUFFER_TOO_SMALL);
+    }
+
+    /* Create the device-mapper task object */
+    if (!(dmt = dm_task_create(DM_DEVICE_CREATE)))
+        RAISE(VIC_FAILED);
+
+    /* Set the device-mapper name */
+    if (!dm_task_set_name(dmt, name))
+        RAISE(VIC_FAILED);
+
+#ifdef TRACE_TARGET
+    printf("DM-NAME: %s\n", name);
+    printf("TARGET: start{%lu} size{%lu} target{%s} params{%s}\n",
+        start, size, "integrity", params);
+#endif
```

Comment:
Eventually this should be integrated with SGX-LKL tracing. We also need to decide what userspace tracing will look like, as this won't be able to use the kernel-level tracing macros.

Reply:
I created an issue to track vicsetup tracing: https://github.com/lsds/sgx-lkl/issues/607. I pressume that **cryptsetup** was not integrated with sgx-lkl tracing either.

Good point about user-space tracing. Can we use syslog?

Reply:
syslog is probably a good idea.  At some point, we'll add hooks to redirect syslog to an external secure audit log.